### ISSUE TITLE
Autodetect SSL-only server and use SSL

### ISSUE THREAD

**Issue by jaraco** [phpmyadmin/phpmyadmin#13436]
Attempting to deploy phpMyAdmin against an Azure hosted MySQL instance, Azure enforces SSL by default. Other deployments following best practices likely do the same.

Late releases of `mysql` client, when connecting to such a server, will automatically connect to this server without issue, apparently auto-detecting the need for SSL and enabling it.

Connecting phpMyAdmin to this server produces the following error:

```
#2001 - SSL Connection is required. Please specify SSL options and retry.
```

Of course, it's possible in the config to specify SSL, but doing so behind a docker container is non-trivial, especially when multiple servers are used.

Ideally, phpMyAdmin could do something similar to what the mysql client does, using SSL if available and enforced at the server. Such a feature would alleviate other open SSL tickets in this project.

**Comment by nijel** [comment]
I think this is bad idea - to establish trusted SSL channel, you need to have the server or CA certificate and this is something what can't be done automatically (as MySQL server certificates usually do not rely on the existing PKI). Just enabling SSL without certificate verification just brings false feeling of using secure connection, while it is not (without certificate verficiation anybody can do mitm attack pretending to be your MySQL server and get access to all data).

**Comment by nijel** [comment]
Thinking more about that, it's no worse than not using SSL connection....

**Comment by emanuelb** [comment]
It's a lot better then no SSL at all as it protect against passive attackers (who only capture the data on the wire) but doesn't protect at all against any MITM attack (which is very hard to detect)
it will be better to add permanent message that say something which reflect:
'The connection to database is encrypted but not verified and thus vulnerable to MITM attacks, please configure the SSL configuration parameters, more information at: ...'
after connection to database over SSL without verification.

**Comment by jaraco** [comment]
> better to add permanent message

If it's important to warn when insecure SSL is used, I suggest such a message should also be posted when the client connects without SSL.

@nijel: I'm looking forward to testing the fix. Thanks for looking into this.

**Comment by jaraco** [comment]
I see I'm currently relying on the Dockerhub container phpmyadmin/phpmyadmin. Any chance someone could release a tagged release with this commit that I could use to confirm the fix? I'm sure I could, but I'm a novice with Docker.

**Comment by emanuelb** [comment]
> If it's important to warn when insecure SSL is used, I suggest such a message should also be posted when the client connects without SSL.

There already a issue for that in:
https://github.com/phpmyadmin/phpmyadmin/issues/12354 "Suggest SSL connection when logging in"

> I see I'm currently relying on the Dockerhub container phpmyadmin/phpmyadmin.

The `edge-4.7` and `edge-4.8` tags use latest daily development releases at:
https://www.phpmyadmin.net/downloads/#devel
see:
https://github.com/phpmyadmin/docker/blob/master/README.md#docker-hub-tags

**Comment by nijel** [comment]
@jaraco The 4.7.3 release is already available as Docker image, you can also use the edge tags as mentioned above.

**Comment by jaraco** [comment]
Hi @nijel: I'm just getting around to testing this. Unfortunately, I'm still getting an error `#9002 - SSL connection is required. Please specify SSL options and retry.` I'm launching the container unmodified only with PMA_HOST defined. Are there other settings that need to be configured to enable SSL to be used (especially if the server requires it)?

**Comment by nijel** [comment]
Apparently you're getting *different* error code for the same message. Argh, I wish MySQL would be consistent in this....

**Comment by jaraco** [comment]
Awesome. I'll grab the nightly tomorrow and try it out.

**Comment by jaraco** [comment]
I tried using the `edge-4.7` tag of the docker image, but best I can tell, the [daily snapshopts](https://www.phpmyadmin.net/downloads/#devel) are stale so I won't be able to test until those snapshots are refreshed or a release is cut.

**Comment by ibennetch** [comment]
Thanks for the report, I've opened a new issue about that and we'll let you know once their fixed. https://github.com/phpmyadmin/phpmyadmin/issues/13820

**Comment by nijel** [comment]
The daily snapshots are now up to date and the docker images are just building.

**Comment by jaraco** [comment]
It's looking better. I'm no longer getting the error, but now after authenticating and being directed to index.php, but that page returns a 500 Internal Server Error with nothing in the payload.

I checked the logs from the container, but there's nothing there in the output. If there's an error message, it's not emitted in the stderr/stdout of the nginx host.

Disabling the SSL requirement on the server restores the proper behavior, so the 500 error does seem relevant to SSL detection.

Can you suggest how I might be able to reveal the underlying failure?

**Comment by jaraco** [comment]
Is there perhaps an environment variable I can set that will let phpMyAdmin render the error message in the browser?

**Comment by nijel** [comment]
@jaraco the PHP logs should be placed in /var/log/php-fpm.log, you can display it using `docker exec phpmyadmin cat /var/log/php-fpm.log`.

**Comment by jaraco** [comment]
Thanks @nijel. I started up the container locally, pointed it at my Azure-hosted SSL-requiring MySQL instance. Attempted to log in and got the 500 response, but then there's nothing in the logs:

```
$ docker exec -it $(docker ps -q) ls -la /var/log/php-fpm.log
-rw-r--r--    1 nobody   nobody           0 Nov 24 15:19 /var/log/php-fpm.log
```

But, by inspecting the running container (thanks for introducing me to `docker exec`), I did find some an error message that appears relevant:

```
$ docker exec $(docker ps -q) cat /var/log/nginx-error.log
2017/11/24 15:20:45 [error] 24#24: *31 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught Error: Call to undefined function PMA\libraries\dbi\_() in /www/libraries/dbi/DBIMysqli.php:160
Stack trace:
#0 /www/libraries/DatabaseInterface.php(2363): PMA\libraries\dbi\DBIMysqli->connect('optout@pixielan...', 'a6926aaec1dc', Array)
#1 /www/libraries/common.inc.php(748): PMA\libraries\DatabaseInterface->connect(256)
#2 /www/index.php(20): require_once('/www/libraries/...')
#3 {main}
  thrown in /www/libraries/dbi/DBIMysqli.php on line 160" while reading response header from upstream, client: 172.17.0.1, server: _, request: "POST /index.php HTTP/1.1", upstream: "fastcgi://unix:/var/run/php/php-fpm.sock:", host: "localhost:8080"
```

Does that indicate why the request is failing?

**Comment by nijel** [comment]
Ah, now I see the problem, will fix it soon.

**Comment by nijel** [comment]
Fixed in f7e6366d016c2624f1690a84bff6dfa6ad4ea178.

**Comment by jaraco** [comment]
The nightly hasn't updated; I'll check again for it in a few days.

**Comment by jaraco** [comment]
tl;dr: It's working now!

----

I'm still not seeing the fix in edge-4.7. I tried edge-4.8 momentarily and that seems to be working, but edge-4.7 is giving me the same error as before. I've pulled the latest image:

```
$ docker pull phpmyadmin/phpmyadmin:edge-4.7
edge-4.7: Pulling from phpmyadmin/phpmyadmin
b1f00a6a160c: Already exists 
ed8452a2cff1: Pull complete 
2467cb692a8f: Pull complete 
579172efe0d1: Pull complete 
c960c0de90ec: Pull complete 
a1470dc71c94: Pull complete 
7b84943c6163: Pull complete 
Digest: sha256:25fe76d47e159252798e227342e5b2d822be4dfdc6d9bc4e70c316dc0cf35a14
Status: Downloaded newer image for phpmyadmin/phpmyadmin:edge-4.7
```

And run the server against that new image:

```
$ docker run -e PMA_HOST=$MYSQL_ADDRESS -p 8080:80 phpmyadmin/phpmyadmin:edge-4.7
2017-12-01 14:03:37,580 CRIT Supervisor running as root (no user in config file)
2017-12-01 14:03:37,580 WARN Included extra file "/etc/supervisor.d/nginx.ini" during parsing
2017-12-01 14:03:37,581 WARN Included extra file "/etc/supervisor.d/php.ini" during parsing
2017-12-01 14:03:37,587 INFO RPC interface 'supervisor' initialized
2017-12-01 14:03:37,588 CRIT Server 'unix_http_server' running without any HTTP authentication checking
2017-12-01 14:03:37,588 INFO supervisord started with pid 1
2017-12-01 14:03:38,592 INFO spawned: 'php-fpm' with pid 20
2017-12-01 14:03:38,595 INFO spawned: 'nginx' with pid 21
2017-12-01 14:03:39,623 INFO success: php-fpm entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2017-12-01 14:03:39,624 INFO success: nginx entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
```

I then log into phpMyAdmin, get the blank page, and the same error appears in the nginx logs:

```
$ docker exec $(docker ps -q) cat /var/log/nginx-error.log
2017/12/01 14:05:52 [error] 25#25: *1 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught Error: Call to undefined function PMA\libraries\dbi\_() in /www/libraries/dbi/DBIMysqli.php:160
Stack trace:
#0 /www/libraries/DatabaseInterface.php(2363): PMA\libraries\dbi\DBIMysqli->connect('user@pix...', '...', Array)
#1 /www/libraries/common.inc.php(748): PMA\libraries\DatabaseInterface->connect(256)
#2 /www/index.php(20): require_once('/www/libraries/...')
#3 {main}
  thrown in /www/libraries/dbi/DBIMysqli.php on line 160" while reading response header from upstream, client: 172.17.0.1, server: _, request: "POST /index.php HTTP/1.1", upstream: "fastcgi://unix:/var/run/php/php-fpm.sock:", host: "localhost:8080"
```

Aha!

I noticed that `edge-4.7` is 5 hours `4.7`. Running simply with `4.7 `is working. I guess edge will catch up soon.
